### PR TITLE
Remove the state machine

### PR DESCRIPTION
**Pull request by typfel**
## background

Initially the sync engine was build around a state machine where the request which would be created were determined by which state it was in. 

*states:*
`Unauthenticated`
`CatchUpPhase1`
`CatchUpPhase2`
`EventProcessing`
`Background`
`BackgroundTask`
`BackgroundFetch`

Each state would have have a list of `transcoders` (login, conversation, ...) which would be asked to do requests to fulfil their task when the state is active. For example the `unauthenticated` state would ask the `LoginTranscoder` and `RegistrationTranscoder` to do requests until the state machine transitioned into another state.

As the sync engine became more complex we ran into situations where we want to do operations which crosses the boundaries of the states. We then introduced another request producing entity which we call `RequestStrategy`, it doesn't live inside a `state` but instead has a set of `status objects` which it query to determine if it should create a request. 

*some example of status objects:*
`AuthenticatedStatus`:  knows if the user is authenticated
`ClientRegistrationStatus`: knows if the user has a valid user client

## what changed

Since It's not good to have two models for generating requests this PR removes the state machine and any business logic contained in it has been moved to new `status objects`. The transcoders which lived inside the state machine has been re-factored into request strategies.

The two new `status objects` are:

* `OperationStatus` :
keeps track of application state like if it's in the foreground or if it's in background performing an audio call or doing a background fetch.

* `SyncStatus` :
Keeps track of if we need to synchronise conversations, connections etc.. 

## dependencies

- [x] https://github.com/wireapp/wire-ios-transport/pull/28
- [x] https://github.com/wireapp/wire-ios-request-strategy/pull/24
- [x] https://github.com/wireapp/wire-ios-message-strategy/pull/59
- [x] https://github.com/wireapp/wire-ios-mocktransport/pull/29

### CODE REVIEW

File: Source/Synchronization/Strategies/TypingStrategy.swift [hunk 1]
```diff
@@ -47,25 +47,25 @@ public struct TypingEvent {
...
     
     fileprivate var typing : ZMTyping!
     fileprivate var conversations : [NSManagedObjectID : Bool] = [:]
     fileprivate var lastSentTypingEvent : TypingEvent?
-    fileprivate let syncContext: NSManagedObjectContext
-    fileprivate unowned var clientRegistrationDelegate: ClientRegistrationDelegate
-    
     fileprivate var tornDown : Bool = false
+
+    @available (*, unavailable)
```
Comment (daehn) [inline Source/Synchronization/Strategies/TypingStrategy.swift]: 👍

File: Source/Synchronization/Transcoders/ZMCallStateLogger.m [hunk 2]
```diff
@@ -21,26 +21,26 @@
...
 @implementation ZMCallStateLogger
 
 
-- (instancetype)initWithFlowSync:(ZMFlowSync *)flowSync
+- (instancetype)initWithFlowSync:(ZMCallFlowRequestStrategy *)callFlowRequestStrategy
 {
     self = [super init];
     if (self) {
-        self.flowSync= flowSync;
+        self.callFlowRequestStrategy= callFlowRequestStrategy;
```
Comment (daehn) [inline Source/Synchronization/Transcoders/ZMCallStateLogger.m]: Space missing

File: Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m [hunk 3]
```diff
@@ -1097,7 +1090,8 @@ - (void)testThatItAddsConversationsFromAResponse
     [self.syncMOC performGroupedBlockAndWait:^{
         // then
         [self checkThatThereAreConversationsForAllRawConversations:rawConversations failureRecorder:NewFailureRecorder()];
-        XCTAssertTrue(self.sut.isSlowSyncDone);
+//        XCTAssertTrue(self.sut.isSlowSyncDone);
```
Comment (daehn) [inline Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m]: Can be removed?

File: Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m [hunk 4]
```diff
@@ -1089,6 +1090,8 @@ - (void)testThatItAddsConversationsFromAResponse
     [self.syncMOC performGroupedBlockAndWait:^{
         // then
         [self checkThatThereAreConversationsForAllRawConversations:rawConversations failureRecorder:NewFailureRecorder()];
+//        XCTAssertTrue(self.sut.isSlowSyncDone);
```
Comment (typfel) [inline Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m]: 🔥

**Comment by mikeger** [comment]
🔥 🚒

File: Source/Synchronization/ZMOperationLoop.m [hunk 5]
```diff
@@ -366,3 +356,11 @@ - (void)pushChannelDidOpen:(ZMPushChannelConnection *)channel withResponse:(NSHT
 
 @end
 
+@implementation ZMOperationLoop (OperationStatus)
+
+- (void)operationStatusWithDidChangeState:(enum SyncEngineOperationState)state
```
Comment (daehn) [inline Source/Synchronization/ZMOperationLoop.m]: The naming confuses me a bit.

Comment (typfel) [inline Source/Synchronization/ZMOperationLoop.m]: Yep doesn't make sense.

`operationStatusDidChangeState ` or you have a better suggestion?

Comment (daehn) [inline Source/Synchronization/ZMOperationLoop.m]: Sounds good 👍

Comment (typfel) [inline Source/Synchronization/ZMOperationLoop.m]: Swift to objc name mangling

File: Source/UserSession/OperationStatus.swift [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            NotificationCenter.default.post(name: OperationStatusChangedNotification, object: self)
+        }
+    }
+    
+    public func startBackgroundFetch(withCompletionHandler completionHandler: @escaping BackgroundFetchHandler) {
+        startBackgroundFetch(timeout: 30.0, withCompletionHandler: completionHandler)
+    }
+    
+    public func startBackgroundFetch(timeout: TimeInterval, withCompletionHandler completionHandler: @escaping BackgroundFetchHandler) {
+        guard backgroundFetchHandler == nil else {
```
Comment (daehn) [inline Source/UserSession/OperationStatus.swift]: Shouldn't this `guard` ensure that the handler is not `nil`?

File: Source/UserSession/OperationStatus.swift [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        backgroundFetchTimer = Timer.scheduledTimer(timeInterval: timeout, target: self, selector: #selector(backgroundFetchTimeout), userInfo: nil, repeats: false)
+        RequestAvailableNotification.notifyNewRequestsAvailable(self)
+    }
+    
+    public func startBackgroundTask(withCompletionHandler completionHandler: @escaping BackgroundTaskHandler) {
+        startBackgroundTask(timeout: 30.0, withCompletionHandler: completionHandler)
+    }
+    
+    public func startBackgroundTask(timeout: TimeInterval, withCompletionHandler completionHandler: @escaping BackgroundTaskHandler) {
+        guard backgroundTaskHandler == nil, isInBackground else {
```
Comment (daehn) [inline Source/UserSession/OperationStatus.swift]: Same here.

File: Source/UserSession/OperationStatus.swift (hunk 6 above)
Comment (daehn) [inline Source/UserSession/OperationStatus.swift]: Ah just saw this is the local one.

Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: No, this guard ensures that we only have one fetch operation running at any time. if `backgroundFetchHandler ` is not nil there's already a fetch operation ongoing.

File: Source/UserSession/OperationStatus.swift (hunk 7 above)
Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: See other answer. However the plan is to re-factor this code to support multiple concurrent background tasks.

File: Source/UserSession/ZMClientRegistrationStatus.h [hunk 8]
```diff
@@ -55,26 +55,22 @@ typedef NS_ENUM(NSUInteger, ZMClientRegistrationPhase) {
...
-@protocol ZMClientClientRegistrationStatusProvider <NSObject>
 
-/// Whether the current client is ready to use
-@property (nonatomic, readonly) BOOL clientIsReadyForRequests;
 
-/// Notify that the current client was deleted remotely
-- (void)didDetectCurrentClientDeletion;
-
-@end
+//@protocol ZMClientClientRegistrationStatusProvider <NSObject>
```
Comment (daehn) [inline Source/UserSession/ZMClientRegistrationStatus.h]: Can be removed?

File: Tests/Source/Integration/APNSTests.m [hunk 9]
```diff
@@ -72,7 +74,8 @@ - (void)testThatAConversationIsCreatedFromAnAPNS
     NSUInteger oldCount = conversationsList.count;
     
     if(useAPNS) {
-        [self.mockTransportSession closePushChannelAndRemoveConsumer]; // do not use websocket
+        self.mockTransportSession.pushChannel.keepOpen = NO;
+//        [self.mockTransportSession closePushChannelAndRemoveConsumer]; // do not use websocket
```
Comment (daehn) [inline Tests/Source/Integration/APNSTests.m]: Can be removed?

File: Tests/Source/Integration/CallingTests+PushNotifications.m [hunk 10]
```diff
@@ -209,9 +206,10 @@ - (void)testThatItJoinsAVideoCallFromAPushNotification_DuringSyncState
                                                                     sequence:@1
                                                                      session:@"session2"];
 
-    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
-        [session simulatePushChannelClosed];
-    }];
+    // TODO jacob
+//    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
+//        [session simulatePushChannelClosed];
```
Comment (daehn) [inline Tests/Source/Integration/CallingTests+PushNotifications.m]: Can be removed?

File: Tests/Source/Integration/CallingTests+PushNotifications.m [hunk 11]
```diff
@@ -249,7 +248,7 @@ - (void)testThatItJoinsAVideoCallFromAPushNotification_DuringSyncState
         // then
         XCTAssertFalse([self lastRequestContainsSelfStateJoined]);
         XCTAssertTrue(self.userSession.didStartInitialSync);
-        XCTAssertTrue(self.userSession.isPerformingSync);
+//        XCTAssertTrue(self.userSession.isPerformingSync); // TODO jacob
```
Comment (daehn) [inline Tests/Source/Integration/CallingTests+PushNotifications.m]: Can this one be commented-in again?

File: Tests/Source/Integration/EmailRegistrationTests.m [hunk 12]
```diff
@@ -299,7 +299,7 @@ - (void)testThatWhenRegisteringAndNeedToWaitForEmailValidationWeCanCancelTheWait
     
     // when
     [self.userSession registerSelfUser:user];
-    WaitForAllGroupsToBeEmpty(1);
+    NSLog(@"## wait for registerSelfUser");
```
Comment (daehn) [inline Tests/Source/Integration/EmailRegistrationTests.m]: Would be nice to remove the logs.

File: Tests/Source/Synchronization/EventDecoderTests.swift [hunk 13]
```diff
@@ -262,6 +262,7 @@ extension EventDecoderTest {
     }
     
     override func tearDown() {
+        EventDecoder.testingBatchSize = nil
```
Comment (daehn) [inline Tests/Source/Synchronization/EventDecoderTests.swift]: `...Size = nil` sounds odd, but I assume it's an Optional?

File: Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m [hunk 14]
```diff
@@ -2113,47 +2118,6 @@ - (void)testThatItDoesAppendsNewConversationIfRegisteredDevice;
     XCTAssertEqual(messages.count, 1u);
 }
 
-- (void)testThatItDoesNotAppendNewConversationIfNewDevice {
```
Comment (daehn) [inline Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m]: Why did this go?

File: Tests/Source/Synchronization/ZMOperationLoopTests.m [hunk 15]
```diff
@@ -125,9 +125,9 @@ - (void)testThatItInitializesThePushChannel
     
     // given
     self.transportSession = [OCMockObject niceMockForClass:[ZMTransportSession class]];
-    [[self.transportSession stub] closePushChannelAndRemoveConsumer];
+//    [[self.transportSession stub] closePushChannelAndRemoveConsumer];
```
Comment (daehn) [inline Tests/Source/Synchronization/ZMOperationLoopTests.m]: Can be removed?

File: Tests/Source/Synchronization/ZMSyncStrategyTests.m [hunk 16]
```diff
@@ -90,93 +89,82 @@ - (void)setUp
...
+    [[[self.applicationStatusDirectoryMock stub] andReturn:self.operationStatusMock] operationStatus];
+    [(ZMApplicationStatusDirectory *)[[self.applicationStatusDirectoryMock stub] andReturn:self.userProfileImageUpdateStatus] userProfileImageUpdateStatus];
+    
+    id userTranscoder = [OCMockObject niceMockForClass:ZMUserTranscoder.class];
     [[[[userTranscoder expect] andReturn:userTranscoder] classMethod] alloc];
-    (void) [[[userTranscoder expect] andReturn:userTranscoder] initWithManagedObjectContext:self.syncMOC];
+    (void) [[[userTranscoder expect] andReturn:userTranscoder] initWithManagedObjectContext:self.syncMOC applicationStatus:OCMOCK_ANY syncStatus:OCMOCK_ANY];
 
-    self.conversationTranscoder = [OCMockObject mockForClass:ZMConversationTranscoder.class];
+    self.conversationTranscoder = [OCMockObject niceMockForClass:ZMConversationTranscoder.class];
```
Comment (daehn) [inline Tests/Source/Synchronization/ZMSyncStrategyTests.m]: Are we sure we want to use nice mocks here?

File: Tests/Source/UserSession/SyncStatusTests.swift [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+        sut.didFinish(.fetchingConversations)
+        // then
+        XCTAssertNil(uiMOC.zm_lastNotificationID)
+        // when
+        sut.didFinish(.fetchingUsers)
+        // then
+        XCTAssertNotNil(uiMOC.zm_lastNotificationID)
+    }
+    
+    func testThatItDoesNotSetTheLastNotificationIDIfItHasNone(){
```
Comment (daehn) [inline Tests/Source/UserSession/SyncStatusTests.swift]: Naming of the test and what it does seem different to me.

File: Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m (hunk 14 above)
Comment (typfel) [inline Tests/Source/Synchronization/Transcoders/ZMConversationTranscoderTests.m]: We no longer check if it's a "new device" when adding a NewConversationMessage, so the test is not valid any longer.

File: Tests/Source/UserSession/SyncStatusTests.swift (hunk 17 above)
Comment (typfel) [inline Tests/Source/UserSession/SyncStatusTests.swift]: True, I guess was it's testing is that it doesn't clear `zm_lastNotificationID` if didn't receive a new one during the sync. See `persistLastUpdateEventID`

File: Tests/Source/Synchronization/EventDecoderTests.swift (hunk 13 above)
Comment (typfel) [inline Tests/Source/Synchronization/EventDecoderTests.swift]: it's optional.

File: Source/Synchronization/ApplicationStatusDirectory.swift [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+            managedObjectContext.performGroupedBlock {
+                if let callInProgress = notification.userInfo?[CallStateObserver.CallInProgressKey] as? Bool {
+                    self?.operationStatus.hasOngoingCall = callInProgress
+                }
+            }
+        }
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
```
Comment (vytis) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: You need to pas the token returned from `addObserver(forName:object:queue:using)` instead of self

File: Source/Synchronization/ApplicationStatusDirectory.swift [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+    public var deliveryConfirmation: DeliveryConfirmationDelegate {
+        return apnsConfirmationStatus
+    }
+    
+    public var clientRegistrationDelegate: ClientRegistrationDelegate {
+        return clientRegistrationStatus
+    }
+    
+    public var operationState: OperationState {
+        get {
```
Comment (vytis) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: Simply return things here

File: Source/Synchronization/ApplicationStatusDirectory.swift [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+            case .foreground:
+                return .foreground
+            case .background, .backgroundCall, .backgroundFetch, .backgroundTask:
+                return .background
+            }
+        }
+    }
+    
+    public var synchronizationState: SynchronizationState {
+        get {
```
Comment (vytis) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: And here

File: Source/UserSession/SyncStatus.swift [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+@objc public enum SyncPhase : Int {
+    case fetchingLastUpdateEventID, fetchingConnections, fetchingConversations, fetchingUsers, fetchingMissedEvents, done
```
Comment (vytis) [inline Source/UserSession/SyncStatus.swift]: Maybe list them on separate lines for better readability

File: Tests/Source/Calling/CallStateObserverTests.swift [hunk 22]
```diff
@@ -50,10 +51,12 @@ class CallStateObserverTests : MessagingTest {
             conversation.internalAddParticipants(Set<ZMUser>(arrayLiteral:receiver), isAuthoritative: true)
             
             self.conversation = conversation
+
```
Comment (vytis) [inline Tests/Source/Calling/CallStateObserverTests.swift]: While we are at it, try to nil all the properties in `tearDown`

File: Tests/Source/Calling/ZMCallFlowRequestStrategyTests.m [hunk 23]
```diff
@@ -77,16 +79,15 @@ - (void)tearDown
     [self.internalFlowManager stopMocking];
```
Comment (vytis) [inline Tests/Source/Calling/ZMCallFlowRequestStrategyTests.m]: set self.internalFlowManager and self.onDemandFlowManager  to nil

File: Tests/Source/UserSession/ApplicationStatusDirectoryTests.swift [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+    var sut : ApplicationStatusDirectory!
+    
+    override func setUp() {
+        super.setUp()
+        
+        let cookieStorage = ZMPersistentCookieStorage()
+        let cookie = ZMCookie(managedObjectContext: syncMOC, cookieStorage: cookieStorage)
+        let mockApplication = ApplicationMock()
+        
+        sut = ApplicationStatusDirectory(withManagedObjectContext: syncMOC, cookie: cookie!, requestCancellation: self, application: mockApplication, syncStateDelegate: self)
```
Comment (vytis) [inline Tests/Source/UserSession/ApplicationStatusDirectoryTests.swift]: add corresponding `tearDown` as well

File: Tests/Source/UserSession/OperationStatusTests.swift [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+
+class OperationStatusTests : MessagingTest {
+    
+    fileprivate var sut : OperationStatus!
+    
+    
+    override func setUp() {
+        super.setUp()
+        
+        sut = OperationStatus()
```
Comment (vytis) [inline Tests/Source/UserSession/OperationStatusTests.swift]: And here

File: Tests/Source/UserSession/SyncStatusTests.swift [hunk 26]
```diff
@@ -0,0 +1,398 @@
...
+    var mockSyncDelegate : MockSyncStateDelegate!
+    override func setUp() {
+        super.setUp()
+        mockSyncDelegate = MockSyncStateDelegate()
+        sut = SyncStatus(managedObjectContext: uiMOC, syncStateDelegate: mockSyncDelegate)
+    }
+    
+    override func tearDown() {
+        self.uiMOC.zm_lastNotificationID = nil
+        super.tearDown()
```
Comment (vytis) [inline Tests/Source/UserSession/SyncStatusTests.swift]: nil mockSyncDelegate and sut

File: Tests/Source/UserSession/ZMUserSessionTestsBase.m [hunk 27]
```diff
@@ -109,6 +128,12 @@ - (void)setUp
 
 - (void)tearDown
 {
+    [self.clientRegistrationStatus tearDown];
```
Comment (vytis) [inline Tests/Source/UserSession/ZMUserSessionTestsBase.m]: Would be good to recheck if everything is niled here as well

File: Source/Synchronization/ApplicationStatusDirectory.swift (hunk 18 above)
Comment (typfel) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: Good catch!

File: Source/Calling/WireCallCenter.swift [hunk 28]
```diff
@@ -259,10 +259,14 @@ public class WireCallCenter : NSObject {
     // Returns conversations with a non idle call state
     public class func nonIdleCallConversations(inUserSession userSession: ZMUserSession) -> [ZMConversation] {
         var nonIdleConversations : Set<ZMConversation> = Set()
-        
-        let conversationsV3 = WireCallCenterV3.nonIdleCalls.flatMap({ (key: UUID, value: CallState) -> ZMConversation? in
-            return ZMConversation(remoteID: key, createIfNeeded: false, in: userSession.managedObjectContext)
-        })
+
+        if let callCenter = WireCallCenterV3.activeInstance {
```
Comment (basine) [inline Source/Calling/WireCallCenter.swift]: I would have split that into a separate PR, since it's independent from the refactoring of the syncMachine.

Comment (typfel) [inline Source/Calling/WireCallCenter.swift]: I needed to do some small changes to the WireCallCenterV3 so that I could write some tests for how the OperationState knows about ongoing calls.

File: Source/Synchronization/Transcoders/Helper/ZMSimpleListRequestPaginator.m [hunk 29]
```diff
@@ -148,6 +149,7 @@ - (void)updateStateWithResponse:(ZMTransportResponse *)response
 - (void)resetFetching
 {
     self.hasMoreToFetch = YES;
+    self.lastResetFetchDate = [NSDate date];
```
Comment (basine) [inline Source/Synchronization/Transcoders/Helper/ZMSimpleListRequestPaginator.m]: What is this needed for?

Comment (typfel) [inline Source/Synchronization/Transcoders/Helper/ZMSimpleListRequestPaginator.m]: It's compared to the date of when the websocket was established to determine if we need to re-fetch notifications stream again.

File: Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m [hunk 30]
```diff
@@ -219,42 +252,43 @@ - (void)processEvents:(NSArray<ZMUpdateEvent *> *)events
...
-    BOOL inProgress = self.listPaginator.status == ZMSingleRequestInProgress;
-
-    // We want to create a new request if we are either currently fetching the paginated stream
-    // or if we have a new notification ID that rewuires a pingback.
-    BOOL shouldCreateRequest = inProgress || hasNewNotification;
-
-    if (fetchingStream && shouldCreateRequest) {
+    BOOL fetchingStream = self.isFetchingStreamForAPNS || self.isFetchingStreamInBackground || self.isSyncing;
+    
+    // If we receive an APNS while fetching the notification stream we cancel the previous request
```
Comment (basine) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: > // If we receive an APNS while fetching the notification stream we cancel the previous request and start another one.

This might be a source for one of our "request loops"

File: Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m [hunk 31]
```diff
@@ -269,29 +303,55 @@ - (ZMTransportRequest *)nextRequest
...
+    }
+    
     NSUUID *latestEventId = [self processUpdateEventsAndReturnLastNotificationIDFromPayload:response.payload syncStrategy:self.syncStrategy];
     if (latestEventId != nil) {
-        self.lastUpdateEventID = latestEventId;
+        if (response.HTTPStatus == 404 && self.isSyncing) {
+            // If we fail during quick sync we need to re-enter slow sync and should not store the lastUpdateEventID until after the slowSync has been completed
+            // Otherwise, if the device crashes or is restarted during slow sync, we lose the information that we need to perform a slow sync
+            [syncStatus updateLastUpdateEventIDWithEventID:latestEventId];
+            // TODO Sabine: What happens when we receive a 404 when we are fetching the notification for a push notification? In theory we would have to enter slow sync as well or at least not store the lastUpdateEventID until the next proper sync in the foreground
```
Comment (basine) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: I think this should now be a TODO Jacob :D

File: Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m [hunk 32]
```diff
@@ -269,29 +303,55 @@ - (ZMTransportRequest *)nextRequest
...
+    if (response.result == ZMTransportResponseStatusPermanentError && self.isSyncing){
+        [syncStatus failCurrentSyncPhase];
+    }
+    
+    if (!self.listPaginator.hasMoreToFetch && self.lastUpdateEventID != nil && self.isSyncing) {
+        
+        // The fetch of the notification stream was initiated after the push channel was established
+        // so we must restart the fetching to be sure that we haven't missed any notifications.
+        if (syncStatus.pushChannelEstablishedDate.timeIntervalSinceReferenceDate < self.listPaginator.lastResetFetchDate.timeIntervalSinceReferenceDate) {
+            [syncStatus finishCurrentSyncPhase];
```
Comment (basine) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: ah! `lastResetFetchDate` makes sense now

File: Source/Synchronization/ZMCallStateRequestStrategy.m [hunk 33]
```diff
@@ -234,9 +221,17 @@ - (void)checkForOngoingCalls;
     }];
 }
 
-- (NSArray *)requestGenerators;
+- (ZMTransportRequest *)nextRequestIfAllowed
 {
-    return @[self.downstreamSync, self.upstreamSync];
+    for (id<ZMRequestGenerator> requestGenerator in @[self.downstreamSync, self.upstreamSync]) {
+        ZMTransportRequest *request = [requestGenerator nextRequest];
```
Comment (basine) [inline Source/Synchronization/ZMCallStateRequestStrategy.m]: For the future: we have a `firstNonNilReturnedFromSelector` on NSArray.

File: Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m (hunk 30 above)
Comment (typfel) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: Right, was actually wondering about why we restart on each APNS. Maybe it's more efficient to let the previous request finish.

File: Source/Synchronization/ZMCallStateRequestStrategy.m (hunk 33 above)
Comment (typfel) [inline Source/Synchronization/ZMCallStateRequestStrategy.m]: Ah we also have `nextRequest` on an array of `ZMRequestGenerator`.

File: Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m (hunk 31 above)
Comment (basine) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: @typfel did you think about my todo here?

Comment (typfel) [inline Source/Synchronization/Transcoders/ZMMissingUpdateEventsTranscoder.m]: No, not very much. So this scenario happens if you get an APNS and you have synced in over 3 weeks?

**Comment by daehn** [comment]
@typfel 🔥 conflicting
